File: frontend/src/api/dashboardSummary.ts
import apiClient from './client'

// UI-3 operational dashboard aggregate — mirrors the typed backend contract
// (app/api/schemas/dashboard_summary.py). The frontend only LABELS these values; it never
// re-derives an outcome, coverage, or tier — the backend owns every decision.

export interface DashScans {
  total: number // = completed + completed_with_limitations + failed (cancelled tracked separately)
  completed: number
  completed_with_limitations: number
  failed: number
  cancelled: number
}

export interface DashEndpoints {
  selected: number // = verified + not_verifiable + failed + skipped
  verified: number
  not_verifiable: number
  failed: number
  skipped: number
  coverage_percent: number | null // null when selected == 0 (never 100)
}

export interface DashFindings {
  auto_closed: number
  reopened: number
  awaiting_confirmation: number
}

export interface DashTierBlock {
  scans: DashScans
  endpoints: DashEndpoints
  findings: DashFindings
}

export interface OperationalDashboard {
  schema_version: number
  period_days: number
  from: string // UTC ISO
  to: string // UTC ISO
  scans: DashScans
  endpoints: DashEndpoints
  findings: DashFindings
  by_tier: Record<string, DashTierBlock> // always keys "1" and "2"
}

export type DashboardPeriod = 7 | 30 | 90

export const dashboardSummaryApi = {
  async getOperational(tenantId: number, days: DashboardPeriod): Promise<OperationalDashboard> {
    const response = await apiClient.get<OperationalDashboard>(
      `/api/v1/tenants/${tenantId}/dashboard/operational-summary`,
      { params: { days } },
    )
    return response.data
  },
}

File: app/api/schemas/dashboard_summary.py
"""Typed response contract for the UI-3 operational dashboard aggregate (U3-1).

Closed + OpenAPI-visible + ``extra="forbid"`` so the frontend can generate types, and no accidental
or sensitive field can slip into the payload. All values are integers/floats + UTC ISO strings.
"""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class ScansBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    total: int  # = completed + completed_with_limitations + failed (cancelled tracked separately)
    completed: int
    completed_with_limitations: int
    failed: int
    cancelled: int


class EndpointsBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    selected: int  # = verified + not_verifiable + failed + skipped
    verified: int
    not_verifiable: int
    failed: int
    skipped: int
    coverage_percent: Optional[float]  # null when selected == 0 (never 100)


class FindingsBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    auto_closed: int  # distinct findings auto-closed in the window
    reopened: int  # distinct findings reopened in the window
    awaiting_confirmation: int  # current open findings mid-streak (below the close threshold)


class TierBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scans: ScansBlock
    endpoints: EndpointsBlock
    findings: FindingsBlock


class DashboardOperationalSummaryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int
    period_days: int
    window_from: str = Field(alias="from")  # UTC ISO
    window_to: str = Field(alias="to")  # UTC ISO
    scans: ScansBlock
    endpoints: EndpointsBlock
    findings: FindingsBlock
    by_tier: dict[str, TierBlock]  # always keys "1" and "2"
